### PR TITLE
Add UseSumOfInsteadOfFlatMapSize rule

### LINKED ISSUE #5399: New rule: UseSumInsteadOfFlatMapSize
## Context / Expected Behavior of the rule
Similar to `UseAnyOrNoneInsteadOfFind`, it would recognize cases when we are doing `flatMap` instead of calculating the end result directly.

```kotlin
val size = lambdaExpression.collectDescendantsOfType<KtBlockExpression>().flatMap { it.statements }.size
```
[could be](https://github.com/detekt/detekt/pull/5397/files#r991027678)
```kotlin
val size = lambdaExpression.collectDescendantsOfType<KtBlockExpression>().sumOf { it.statements.size }
```
which is simpler and more performant.

I've seen this flatMap + size/isEmpty pattern used a lot previously, even by myself, but we can do better!

So far I can only think of
 * `.flatMap{}.size` -> `.sumOf`
 * `.flatMap{}.isEmpty()` -> could be `.sumOf`, but can be even shortcut with `.none{it.isNotEmpty()}`
 * `.flatMap{}.isNotEmpty()` -> could be `.sumOf`, but can be shortcut with `.any{it.isNotEmpty()}`

There might be other cases too, I think the first one would be good enough to flag, as it has a clearly better replacement. The last 2 might be too complex for the potentially micro-optimization.

### PR DESCRIPTION
**Pull request by t-kameyama**
Fixes #5399

### CODE REVIEW

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseSumOfInsteadOfFlatMapSizeSpec.kt [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    @DisplayName("Reports flatMap and count on Set")
+    fun reportFlatMapAndCountOnSet() {
+        val code = """
+            fun test(set: Set<Bar>) {
+                set.flatMap { it.bar }.count()
+            }
+            class Bar(val bar: Set<Int>)
+        """.trimIndent()
+        val actual = subject.compileAndLintWithContext(env, code)
+        assertThat(actual).hasSize(1)
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseSumOfInsteadOfFlatMapSizeSpec.kt:107]: Could you add some tests to ensure that this rule doesn't flag all the `flatMap` calls?

**Review by BraisGabin — COMMENTED** [review]
Thank you, as always, for your valuable contributions :)

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseSumOfInsteadOfFlatMapSizeSpec.kt (hunk 1 above)
Comment (t-kameyama) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseSumOfInsteadOfFlatMapSizeSpec.kt:107]: dcb4891

**Review by BraisGabin — APPROVED** [review]

**Review by schalkms — APPROVED** [review]

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseSumOfInsteadOfFlatMapSizeSpec.kt [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+import org.junit.jupiter.api.DisplayName
+import org.junit.jupiter.api.Test
+
+@KotlinCoreEnvironmentTest
+class UseSumOfInsteadOfFlatMapSizeSpec(val env: KotlinCoreEnvironment) {
+    val subject = UseSumOfInsteadOfFlatMapSize()
+
+    @Test
+    @DisplayName("Reports flatMap and size")
+    fun reportFlatMapAndSize() {
```
Comment (cortinico) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseSumOfInsteadOfFlatMapSizeSpec.kt:16]: What's the reason why this is not:
```kotlin
fun `reports flatMap and size`() {
```

It seems like this goes a bit against the rest of the codebase convention.

Comment (t-kameyama) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseSumOfInsteadOfFlatMapSizeSpec.kt:16]: https://github.com/detekt/detekt/pull/5416

**Pull request merged**